File: forest_contest/03/03-005-3.rb
# ラッキーナンバー (paizaランク B 相当)
# https://paiza.jp/works/mondai/forest_contest_003/forest_contest_003__b_lucky_number

# 入出力例
INPUT1 = <<"EOS"
4
444
222
333
666
EOS
OUTPUT1 = <<"EOS"
333 444
EOS

INPUT2 = <<"EOS"
4
111
222
333
444
EOS
OUTPUT2 = <<"EOS"
multiple answers
EOS

INPUT3 = <<"EOS"
4
800
125
310
602
EOS
OUTPUT3 = <<"EOS"
no answer
EOS

# ruby 解答例3
BINGO = 7
# 入力
_, *a = INPUT0.split.map(&:to_i)
# BINGO 以下の数字だけ調べる
a.select! { |x| x <= BINGO }

# dpテーブル初期化
dp = Array.new(BINGO + 1) { [] }
# 数字の組み合わせで作れる数を記録していく
a.each do |num|
  p num
  BINGO.downto(num) do |i|
    next if dp[i - num].empty?
    dp[i - num].each do |ary|
      dp[i] << ary + [num]
    end
  end
  # BINGO が複数ならループを抜ける
  break if dp[BINGO].length > 1
  # 単体で作れる数
  dp[num] << [num]
end

# 出力
puts case dp[BINGO].length
  when 0
    # BINGO が作れない場合
    "no answer"
  when 1
    # BINGO の組み合わせが 1 個
    dp[BINGO][0].sort.join(" ")
  else
    # BINGO の組み合わせが 2 個以上
    "multiple answers"
  end

File: forest_contest/03/03-005-3.py
# ラッキーナンバー (paizaランク B 相当)
# https://paiza.jp/works/mondai/forest_contest_003/forest_contest_003__b_lucky_number

# 入出力例
INPUT1 = """\
4
444
222
333
666"""
OUTPUT1 = """\
333 444"""

INPUT2 = """\
4
111
222
333
444"""
OUTPUT2 = """\
multiple answers"""

INPUT3 = """\
4
800
125
310
602"""
OUTPUT3 = """\
no answer"""

INPUT0 = """\
4
1
2
3
4"""

# python3 解答例3
BINGO = 777
# 入力
_, *a = [int(x) for x in open(0).read().split()]

# BINGO 以下の数字だけ調べる
a = [x for x in a if x <= BINGO]
n = len(a)

# dpテーブル初期化
dp = [[] for _ in range(BINGO+1)]
# 数字の組み合わせで作れる数を記録していく
for num in a:
    for i in range(BINGO, num-1, -1):
        if len(dp[i - num]) == 0:
            continue
        for ary in dp[i - num]:
            dp[i].append(ary + [num])

    # BINGO が複数ならループを抜ける
    if len(dp[BINGO]) > 1:
        break
    # 単体で作れる数
    dp[num].append([num])

# 出力
if len(dp[BINGO]) == 0:
    # BINGO が作れない場合
    res = "no answer"
elif len(dp[BINGO]) == 1:
    # BINGO の組み合わせが 1 個
    res = " ".join(map(str, sorted(dp[BINGO][0])))
else:
    # BINGO の組み合わせが 2 個以上
    res = "multiple answers"
print(res)